proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,233 @@
+Types that can be immediately determined to not be intangible (that is, types that
+cannot be a resource type or cannot contain a resource type) are types like booleans,
+int, float, float4, etc. If the resource type is among any numerics or a type that
+cannot be an intangible type, there is only one exception where it can be treated
+as a resource. The only valid register type for such resource types is 'b', which
+adheres to legacy behavior. All other register types applied to such a type will 
+emit an error diagnostic that "the given type '%0' cannot be bound as a resource".
+If the register type is 'b', a warning will be emitted instead, that will be
+treated as an error by default: "register type 'b' used for resource type '%0', 
+which cannot be used as a resource". Below are some examples:
+
+| Code | Diagnostic |
+|-|-|
+| `float f : register(t0)` | "error: 'float' is an invalid resource type for register type 't'" |
+| `float f : register(b0)` | DefaultError "warning: register type 'b' used for resource type 'float', which cannot be used as a resource. Disable with --Wno-disallow-legacy-binding-rules" |
+
+
+
+## Detailed design
+
+All the compiler has to work with is a Decl object. From the Decl object, 
+the ResourceAttr attribute can be fetched, with which the resource 
+class can be determined. Once this is known, we then check if the resource
+type is a user-defined type or if it is a standard resource type. If the 
+resource type is among the standard resource types (e.g., cbuffer, tbuffer,
+RWBuffer, etc.), then the register type is checked for validity. If there is
+an incompatibility between resource class and register type, a diagnostic will
+be emitted. Otherwise, the resource is a user-defined type, and so the type needs
+to be validated to see whether it can properly be assigned to the given register 
+type. If the UDT does not have at least one resource that matches the given register 
+types, then an error will be emitted. The only remaining case is when there
+is no resource attribute, which implies the type cannot be intangible. In this case,
+if the type is strictly numeric, the only acceptable register type is 'b', which 
+is legacy behavior. In this case, a warning will be emitted that will be treated as 
+an error by default, which states that a numeric type is not usable as a resource. 
+If 'c' or 'i' are given as register types, a warning will be emitted stating that
+such register types are legacy behavior, and the resource won't be used. Any other 
+register type will cause an error to be emitted, stating that the given resource
+type cannot be bound as a resource.
+Legacy behavior can be allowed with the --Wno-disallow-legacy-binding-rules
+flag. When this flag is active, and legacy behavior is present, a warning will 
+be emitted instead of an error.
+
+Below is a flowchart that describes the process that is used to determine
+what kind of diagnostic to emit in each case:
+
+```mermaid
+flowchart TD

```

Comment:
Now that the rules are simpler, I wonder if the flowchart is the clearest way to represent this?

Reply:
Tex's feedback has made me aware that this is more complex than previously thought. I think the flowchart should stay now.